[ClangImporter] Use clang getDefaultLanguageStandard

porglezomp opened the pull request:
This moves the logic for selecting the default language standard into Clang, instead of duplicating it.
This is intended to make the reliance on `CLANG_DEFAULT_STD_C(XX)` less fragile, since clang-side refactors caused problems there.

This currently preserves the same standard versions without this change if either of:
- `CLANG_DEFAULT_STD_C(XX)` is set
- ObjC interop is enabled

Questions for reviewers:
- Do we consider it good to have this completely tied to clang? It ensures we'll always be in synch, but it also means that by default we get updated possibly without knowing.
- Is it fine that this bumps the default standards when Objective-C interop is disabled?
- Is it fine that in general, this lets the C vs Objective-C and C++ vs Objective-C++ standard diverge.

Future question, I guess: Should it be possible to explicitly specify the clang importer standard version in some way?

porglezomp committed:
[ClangImporter] Use clang getDefaultLanguageStandard

This moves the logic for selecting the default language standard into
Clang, instead of duplicating it. This is intended to make the reliance
on CLANG_DEFAULT_STD_C(XX) less fragile, since clang-side refactors
caused problems there.

This uses ObjC and ObjCXX to provide the language standards, since that
matched the existing language standard behavior here.

porglezomp commented:
Will add a second commit here that picks the language based on whether Objective-C and C++ interop are available.

compnerd reviewed (commented):
I generally like the idea of querying this from clang.  I can see the concern about the version being floating, however, I think that it might actually be safer - the default version is more likely to be what is tested rather than the explicit version that we select.  It also simplifies the model - consider the question that comes up when we start looking at this, why is it gnu11 that we are targeting rather than std11?  Or perhaps std17?

I am less enthusiastic about the use of the ObjC/ObjC++ for deriving this.  Can we not base that upon whether ObjC interop is enabled or not?  I also question if we should be using the GNU extensions by default.  This is particularly an important point when you consider other platforms.  Perhaps we should also enable MSVC extensions unconditionally?  Using the standard version of the language does mean that we get more portability, but it can also restrict some things which might be problematic.

Subsequently, I have the follow up questions: how much does stuff break if we disable the GNU extensions?  Should we perhaps consider exposing this as a proper flag if we are now tying this to the behaviour of the associated clang because there might be issues that come up with clang version changes and headers from the system.

porglezomp committed:
[ClangImporter] Use the enabled interop to pick the language standard

clang::LangStandard::getDefaultLanguageStandard takes a language that it
uses to select the default standard. Now we properly pick one of C, C++,
ObjC, and ObjC++, depending on which interop features are enabled.

This implicitly upgrades no-ObjC + yes-C++ interop from C++14 to C++17,
matching the default selected by Clang.

porglezomp commented:
Switched to selecting the language based on which interop features are enabled.

Note that this doesn't disable the GNU extensions, `getDefaultLanguageStandard` currently returns `gnu17` for C, `gnu11` for ObjC, `gnucxx17` for C++, and `gnucxx14` for ObjC++.
See the relevant Clang code here: https://github.com/apple/llvm-project/blob/next/clang/lib/Basic/LangStandards.cpp#L60-L81

I don't have much of an opinion on what extensions should be enabled, and I'm not sure who'd be the best person to ask.
With this change specifically I'd like to try to minimize the differences it's introducing.
It *does* change the behavior on platforms without Objective-C interop, since Clang has moved to newer standards for those than we had specified in Swift.

This means the standard revisions used by Swift diverge depending on whether Objective-C interop is enabled, which might not be desirable—but it is the behavior that Clang demonstrates when you're building C vs Objective-C, and C++ vs Objective-C++.

porglezomp commented:
Updated the issue description to have a more relevant set of questions after the second commit.

zoecarver commented:
This LGTM, thank you!

porglezomp commented:
@swift-ci please test

porglezomp commented:
@compnerd what do you think after the change to make it pick the language based on the enabled interop features? Any concerns about upgrading the default standard on Linux and Windows?

porglezomp commented:
@swift-ci please test

compnerd reviewed (approved):
I think that this is a much better approach.  While it doesn't explicitly disable the GNU extensions, it doesn't also explicitly enable them - the control of the extensions is now within clang, which can make the decision based on the target.  Arguably, this is *better* than disabling the extensions explicitly.  Given that clang is given both the language and the host triple, it should be able to enforce that we do not go below the minimum viable standards and can appropriately select the standard, I think that this is good.

porglezomp commented:
@swift-ci please test
